Public/Src/Tools/Execution.Analyzer/Analyzers.Core/XLGPlusPlus/XldbProtobufExtensions.cs

```diff
@@ -558,5 +548,479 @@ public static Xldb.Fingerprint ToFingerprint(this Fingerprint fingerprint)
                 Bytes = Google.Protobuf.ByteString.CopyFrom(fingerprint.ToByteArray())
             };
         }
+
+        /// <nodoc />
+        public static Xldb.DirectoryArtifact ToDirectoryArtifact(this DirectoryArtifact artifact, PathTable pathTable)
+        {
+            return new Xldb.DirectoryArtifact()
+            {
+                IsValid = artifact.IsValid,
+                Path = artifact.Path.ToAbsolutePath(pathTable),
+                PartialSealID = artifact.PartialSealId,
+                IsSharedOpaque = artifact.IsSharedOpaque
+            };
+        }
+
+        /// <nodoc />
+        public static Xldb.PipTable ToPipTable(this PipTable pipTable)
+        {
+            var xldbPipTable = new Xldb.PipTable
+            {
+                IsDisposed = pipTable.IsDisposed,
+                Reads = pipTable.Reads,
+                Writes = pipTable.Writes,
+                Count = pipTable.Count,
+                PageStreamsCount = pipTable.PageStreamsCount,
+                Size = pipTable.Size,
+                WritesMilliseconds = pipTable.WritesMilliseconds,
+                ReadsMilliseconds = pipTable.ReadsMilliseconds,
+                Used = pipTable.Used,
+                Alive = pipTable.Alive

```

Comment:
I don't think these internal implementation details of the PipTable structure are particularly useful to put into the database. Do you even need the PipTable structure at all?

Reply:
It is relatively cheap (to insert and storage wise) and a single entry into the db. The most important things here are the stablekeys and keys that stored since this allows you to iterate over all pips. Thus I think some, but not all of this metadata could be immediately useful, and the others are "nicetohaves" 

---
In reply to: [314443512](https://github.com/microsoft/BuildXL/pull/724#discussion_r314443512) [](ancestors = 314443512)